mcnode/node.go

```diff
@@ -323,3 +343,106 @@ func (node *Node) doShutdown() {
 	node.ds.Close()
 	os.Exit(0)
 }
+
+func (auth *PeerAuth) fromJSON(rmap map[string]interface{}) error {
+	auth.mx.Lock()
+	defer auth.mx.Unlock()
+
+	auth.peers = make(map[p2p_peer.ID][]string)
+	for id58, xrules := range rmap {
+		pid, err := p2p_peer.IDB58Decode(id58)
+		if err != nil {
+			return err
+		}
+
+		xxrules, ok := xrules.([]interface{})
+		if !ok {
+			return BadRuleset
+		}
+
+		rules := make([]string, len(xxrules))
+		for x, xxrule := range xxrules {
+			rule, ok := xxrule.(string)
+			if !ok {
+				return BadRuleset
+			}
+
+			rules[x] = rule
+		}
+
+		auth.peers[pid] = rules
+	}
+
+	return nil
+}
+
+func (auth *PeerAuth) toJSON() map[string]interface{} {
+	auth.mx.Lock()
+	defer auth.mx.Unlock()
+
+	rmap := make(map[string]interface{})
+	for pid, rules := range auth.peers {
+		rmap[pid.Pretty()] = rules
+	}
+
+	return rmap
+}
+
+func (auth *PeerAuth) authorize(pid p2p_peer.ID, nss []string) bool {
+	if len(nss) == 0 {
+		return false
+	}
+
+	auth.mx.Lock()
+	defer auth.mx.Unlock()
+
+	rules := auth.peers[pid]
+	if len(rules) == 0 {
+		return false
+	}
+
+	for _, ns := range nss {
+		if !auth.authorizeAllow(rules, ns) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (auth *PeerAuth) authorizeAllow(rules []string, ns string) bool {
+	for _, rule := range rules {
+		switch {
+		case rule == "*":
+			return true
+
+		case strings.HasSuffix(rule, ".*"):
+			if strings.HasPrefix(ns, rule[:len(rule)-2]) {
```

Comment:
Uhm, I don't think it's worth the trouble.
I kept it simple and used the same wildcard semantics we have in MCQL.
If it proves to be insufficient for some use case, we can generalize to support regexps.